server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -402,6 +407,11 @@ public boolean shouldCache(Query query) {
         persistMetadata(path, indexSettings, shardRouting, null, logger);
         this.useRetentionLeasesInPeerRecovery = replicationTracker.hasAllPeerRecoveryRetentionLeases();
         this.refreshPendingLocationListener = new RefreshPendingLocationListener();
+        if (indexSettings.isSegRepEnabled() == true) {
+            this.checkpointRefreshListener = new CheckpointRefreshListener(this, checkpointPublisher);
```

Comment:
I'd prefer not including the check here for the feature flag.  We know if segrep is off we will get a NoOp version of the publisher from IndexService, we can also add another check inside the publisher if we need to.  If we don't want the listener wired up at all I'd suggest we then wire up a list of additional listeners and pass those from IndexService - [related](https://github.com/opensearch-project/OpenSearch/pull/3285#discussion_r874192885).

Reply:
Yes this check looks redundant, not that useful. If we want additional listeners to be passed from IndexService, then we need to change constructor of IndexShard and pass additonal Listeners instead of checkpointPublisher. By default additonal refreshListeners would be empty but in case of segrep on, we then pass a checkpointrefreshlistener from IndexService to IndexShard. Is this right @mch2  ?

Reply:
You are right.  The only other way to do this is then always wire up the listener and it will be NoOp when segrep is off.
I'm ok with this extra logic for now while the feature is behind a feature flag.

Reply:
I'm not clear on why we need a duplicate `isSegRepEnabled` check here.

If the conditions in `IndexService` are not satisfied, there's no reason to pass a NoOp checkpoint publisher because the notion of checkpoint publishing does not apply at all. The NoOp implementation should only be needed for test cases.

Within the actual code path, the checkpoint publisher object for non-seg-rep cases should be `null`, and `checkpointRefreshListener` should simply check for that for initialization.

Reply:
Thanks @kartg , made changes suggested by you in latest commit.